devdoc/constbuffer_array_splitter_requirements.md

```diff
@@ -58,3 +60,49 @@ This takes the `buffers` from a `CONSTBUFFER_ARRAY_HANDLE` and splits them into
 **SRS_CONSTBUFFER_ARRAY_SPLITTER_42_016: [** `constbuffer_array_splitter_split` shall succeed and return the `split_buffers`. **]**
 
 **SRS_CONSTBUFFER_ARRAY_SPLITTER_42_017: [** If there are any other failures then `constbuffer_array_splitter_split` shall fail and return `NULL`. **]**
+
+### constbuffer_array_splitter_split_to_array_of_array
+
+```c
+MOCKABLE_FUNCTION(, TARRAY(CONSTBUFFER_ARRAY_HANDLE), constbuffer_array_splitter_split_to_array_of_array, CONSTBUFFER_ARRAY_HANDLE, buffers, uint32_t, max_buffer_size);
+```
+
+This takes the `buffers` from a `CONSTBUFFER_ARRAY_HANDLE` and splits them into a new `TARRAY(CONSTBUFFER_ARRAY_HANDLE)` where each `CONSTBUFFER_ARRAY_HANDLE` is of size `max_buffer_size` (or smaller for the final remaining buffer). The caller is responsible for cleaning up the returned `TARRAY(CONSTBUFFER_ARRAY_HANDLE)`.
+
+**SRS_CONSTBUFFER_ARRAY_SPLITTER_07_001: [** If `buffers` is `NULL` then `constbuffer_array_splitter_split_to_array_of_array` shall fail and return `NULL`. **]**
+
+**SRS_CONSTBUFFER_ARRAY_SPLITTER_07_002: [** If `max_buffer_size` is `0` then `constbuffer_array_splitter_split_to_array_of_array` shall fail and return `NULL`. **]**
+
+**SRS_CONSTBUFFER_ARRAY_SPLITTER_07_005: [** `constbuffer_array_splitter_split_to_array_of_array` shall call `constbuffer_array_get_buffer_count` to get the total number of buffers. **]**
+
+**SRS_CONSTBUFFER_ARRAY_SPLITTER_07_008: [** `constbuffer_array_splitter_split_to_array_of_array` shall call `constbuffer_array_get_all_buffers_size` for `buffers` and store the result as `remaining_buffer_size`. **]**
+
+**SRS_CONSTBUFFER_ARRAY_SPLITTER_07_021: [** If `buffer_count` or remaining_buffer_size is `0`, `constbuffer_array_splitter_split_to_array_of_array` shall call `TARRAY_CREATE_WITH_CAPACITY` with size 1.  **]**
```

Comment:
>remaining_buffer_size [](http://example.com/codeflow?start=66&length=21)

remaining_buffer_size sounds like an implementation detail. This is probably because first the code was written and then specs were done :-).

I think the specs should look like:
If the total size for all buffers in `buffers` is 0, then  `constbuffer_array_splitter_split_to_array_of_array` shall call `TARRAY_CREATE_WITH_CAPACITY` with size 1.

Reply:
Actually it should probably be like:
If the total size for all buffers in buffers is 0:
- `constbuffer_array_splitter_split_to_array_of_array` shall call `TARRAY_CREATE_WITH_CAPACITY` with size 1.
- `constbuffer_array_splitter_split_to_array_of_array` shall call `constbuffer_array_create_empty` and store the created const buffer array in the first entry of the TARRAY that was created.
- `constbuffer_array_splitter_split_to_array_of_array` shall return the TARRAY.

Reply:
After reading all the specs here, I think if you rework a little the specs to not have `remaining_buffer_size` (which is internal implementation detail of the code it should be pretty much fine.

Reply:
Yeah, after reading all of this I think the changes should be minor:
- the one mentioned above about the case of 0 buffers
- Normally we should not have remaining_buffer_size here in the spec, which is a detail of how the variable is named in the code (suggest saying "total size of buffers in `buffers`). If you really want to shorten it in the specs you can do something like:

**SRS_CONSTBUFFER_ARRAY_SPLITTER_07_008: [** `constbuffer_array_splitter_split_to_array_of_array` shall call `constbuffer_array_get_all_buffers_size` for `buffers` and to obtain the total size of all buffers in `buffers`. **]** The total size will be reference in subsequent specs as `total_buffer_size`.

Reply:
done